Compiler error from match expression returning optional

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Arch Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 5600X 6-Core Processor
CC version: cc (GCC) 11.1.0

getwd: /tmp
vmodules: /home/rcorre/.vmodules
vroot: /home/rcorre/src/v
vexe: /home/rcorre/src/v/v
vexe mtime: 2021-05-30 17:49:23
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 2abbbcc.17deaca

Git version: git version 2.31.1
Git vroot status: weekly.2021.21-68-g17deacaa
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg example.v`
```v
type Any = int | string

fn ok(s string) Any {
	return match s {
		'foo' {
			Any(1)
		}
		else {
			Any('asdf')
		}
	}
}

fn fails(s string) ?Any {
	return match s {
		'foo' {
			Any(1)
		}
		else {
			Any('asdf')
		}
	}
}

ok('foo')
fails('foo') ?

```

**What did you expect to see?**

A match expression can return an optional type


**What did you see instead?**
```
==================
/tmp/v/example.6930156607422942128.tmp.c:10179: error: '{' expected (got ";")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5358,7 +5358,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					}
 					expr_type := c.expr(stmt.expr)
 					if ret_type == ast.void_type {
-						if node.is_expr
+						if node.is_expr && !node.expected_type.has_flag(.optional)
 							&& c.table.get_type_symbol(node.expected_type).kind == .sum_type {
 							ret_type = node.expected_type
 						} else {

--- a/vlib/v/tests/match_expr_returning_optional_test.v
+++ b/vlib/v/tests/match_expr_returning_optional_test.v
@@ -1,0 +1,36 @@
+type Any = int | string
+
+fn ok(s string) Any {
+	return match s {
+		'foo' {
+			Any(1)
+		}
+		else {
+			Any('asdf')
+		}
+	}
+}
+
+fn fails(s string) ?Any {
+	return match s {
+		'foo' {
+			Any(1)
+		}
+		else {
+			Any('asdf')
+		}
+	}
+}
+
+fn test_match_expr_returning_optional() {
+	ret1 := ok('foo')
+	println(ret1)
+	assert ret1 == Any(1)
+
+	ret2 := fails('foo') or {
+		assert false
+		return
+	}
+	println(ret2)
+	assert ret2 == Any(1)
+}

